Webcam flicker

stephencookdev opened the pull request:
# Problem
Currently when we load the webcam page, we render...

- the fallback (if we're allowing fallbacks)
- the permissions primer ("you need to allow permissions")
- and then the correct camera screen

# Solution

This PR just makes sure to render `null` when we're not sure what the state of "permissions" or camera capabilities are, so as to avoid the page flicker


## Checklist
- [X] Have the CHANGELOG, README, MIGRATION, RELEASE_GUIDELINES docs been updated?
- [n/a] Have new automated tests been implemented?
- [n/a] Have new manual tests been written down?
- [X] Have tests passed locally?
- [n/a] Have any new strings been translated?

rfreitas commented:
Travis automatic deployment: https://staging-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-565-pr-onfido-sdk-ui-onfido.surge.sh

stefaniacardenas commented on src/components/Capture/withCameraDetection.js:
```diff
@@ -2,10 +2,17 @@ import { h } from 'preact'
+// previous value was (but we still actually check, and update the value correctly)
+// If we are incorrect, then the user will simply see a flicker of a fallback,
+// or an empty camera stream, before the correct screen once more
+let previousHasCameraCheck = null
```
What happened if the webcam is disconnected mid-flow?

stephencookdev replied:
We still _actually_ check the webcam support, throughout. We only use this variable as a means of getting a more sane "start" value (i.e. we assume you have a webcam since you did before, but still _actually_ check that you do, and then correct ourselves if we were wrong)

But I'm assuming this idea of a user disconnecting mid-flow is actually fairly rare, in real use-cases (hence we should still allow it to happen, but not optimise for it)

stephencookdev replied:
Although messing about a bit more, it looks like the way that the forwarding to the cross-device flow when no camera support is enabled works, then our caching is fine when `true` --> `false`
But when `false` --> `true`, you're actually stuck with the code always assuming you don't have a webcam. There are ways we can work around this, but it'll end up just causing more hassle than it's worth - so will just remove the idea of caching here at all

stephencookdev commented:
> Test are failing. I think it's due to the fact that when there is no camera and video is requested it fails by showing the uploader instead of the cross device steps that allow you to send the sms

So interestingly the tests were failing because of a mix of the caching logic that I introduced (now removed, after your other comment), and because of a bug in the `withFlowChangeOnDisconnectCamera` HoC (since we don't actually care about _disconnect_, we still want the flow change even if there was _never_ a camera) - so I tightened up that in https://github.com/onfido/onfido-sdk-ui/pull/565/commits/f7fb5e5f630ff9ffe5358967f52e551edc1f139e

rfreitas commented:
Travis automatic deployment: https://staging-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://565-pr-onfido-sdk-ui-onfido.surge.sh

stephencookdev commented:
This should be ready for a re-review btw @stefaniacardenas, whenever you've got time

rfreitas commented:
Travis automatic deployment: https://staging-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://staging-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://staging-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://staging-565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://565-pr-onfido-sdk-ui-onfido.surge.sh

rfreitas commented:
Travis automatic deployment: https://release-565-pr-onfido-sdk-ui-onfido.surge.sh
